commit c49915439a1fdfb97a8b6f8ae5676c18c5ac2eea
Author: Kartik <kartik@bugqore.com>
Date:   Mon Aug 17 00:58:15 2026 +0530

    Keep check_escape error offset within the pattern for \x{ and \o{ (#948)

diff --git a/src/pcre2_compile.c b/src/pcre2_compile.c
index c64fd273..aeda6370 100644
--- a/src/pcre2_compile.c
+++ b/src/pcre2_compile.c
@@ -2020,7 +2020,7 @@ else
     else
       {
       *errorcodeptr = ERR64;
-      goto ESCAPE_FAILED_FORWARD;
+      if (ptr < ptrend) goto ESCAPE_FAILED_FORWARD;
       }
     break;
 
@@ -2107,7 +2107,7 @@ else
         else
           {
           *errorcodeptr = ERR67;
-          goto ESCAPE_FAILED_FORWARD;
+          if (ptr < ptrend) goto ESCAPE_FAILED_FORWARD;
           }
         }   /* End of \x{} processing */
 
diff --git a/testdata/testinput2 b/testdata/testinput2
index 4bc71fcf..a7c1c846 100644
--- a/testdata/testinput2
+++ b/testdata/testinput2
@@ -3987,6 +3987,10 @@
 
 /^A\x{/
 
+/\x{2/
+
+/\o{7/
+
 /[ab]++/B,no_auto_possess
 
 /[^ab]*+/B,no_auto_possess
diff --git a/testdata/testinput5 b/testdata/testinput5
index d47e0c1e..098f4100 100644
--- a/testdata/testinput5
+++ b/testdata/testinput5
@@ -3616,4 +3616,9 @@
     \x{e1}
     \x{c1}
 
+# An unterminated \N{U+ escape at the end of the pattern must
+# report an error offset that lies within the pattern.
+
+/(*UTF)\N{U+2/
+
 # End of testinput5
diff --git a/testdata/testoutput2 b/testdata/testoutput2
index ccb5667f..acb4adf7 100644
--- a/testdata/testoutput2
+++ b/testdata/testoutput2
@@ -13630,6 +13630,14 @@ Failed: error 167 at offset 8: non-hex character in \x{} (closing brace missing?
 Failed: error 178 at offset 5: digits missing after \x or in \x{} or \o{} or \N{U+}
         here: ^A\x{ |<--|
 
+/\x{2/
+Failed: error 167 at offset 4: non-hex character in \x{} (closing brace missing?)
+        here: \x{2 |<--|
+
+/\o{7/
+Failed: error 164 at offset 4: non-octal character in \o{} (closing brace missing?)
+        here: \o{7 |<--|
+
 /[ab]++/B,no_auto_possess
 ------------------------------------------------------------------
         Bra
diff --git a/testdata/testoutput5 b/testdata/testoutput5
index 82b3abe9..0337d6aa 100644
--- a/testdata/testoutput5
+++ b/testdata/testoutput5
@@ -8335,4 +8335,11 @@ Failed: error 150 at offset 5: invalid range in character class
     \x{c1}
  0: \x{c1}
 
+# An unterminated \N{U+ escape at the end of the pattern must
+# report an error offset that lies within the pattern.
+
+/(*UTF)\N{U+2/
+Failed: error 167 at offset 12: non-hex character in \x{} (closing brace missing?)
+        here: ...UTF)\N{U+2 |<--|
+
 # End of testinput5
